src/simplify_algebra.cpp

```diff
@@ -714,6 +716,128 @@ struct find_rsqrt
+            assert(i->outputs().size() == 1);
+            auto cont = i->outputs().front();
+            assert(cont->outputs().size() == 1);
+            return cont->outputs().front();
+        });
+
+        // all outputs are reshape and of the same shape
+        auto dims = any_cast<op::reshape>(rsp->get_operator()).dims;
+        if(!std::all_of(vec_rsp.begin(), vec_rsp.end(), [&](auto i) {
+               return (i->name() == "reshape") and
+                      (any_cast<op::reshape>(i->get_operator()).dims == dims);
+           }))
+        {
+            return;
+        }
+
+        auto axis    = any_cast<op::slice>(slc->get_operator()).axes[0];
+        auto in_lens = input->get_shape().lens();
+        if(!std::equal(in_lens.begin(), in_lens.begin() + axis, dims.begin(), dims.begin() + axis))
+        {
+            return;
+        }
+
+        // calculate reshape output shape
+        auto tmp_lens = vec_rsp.front()->get_shape().lens();
+        std::vector<int64_t> rsp_lens(tmp_lens.begin(), tmp_lens.end());
+        int64_t dim_size = rsp_lens[axis];
+        rsp_lens[axis] *= vec_rsp.size();
+
+        // insert the reshape instruction
+        auto rsp_ins = p.insert_instruction(std::next(input), op::reshape{rsp_lens}, input);
+
+        // replace the original reshape with slice
+        int64_t i = 0;
+        for(auto in : vec_rsp)
+        {
+            p.replace_instruction(
+                in, op::slice{{axis}, {i * dim_size}, {(i + 1) * dim_size}}, rsp_ins);
+            ++i;
+        }
+    }
+};
+
+struct find_split_transpose
+{
+    auto matcher() const
+    {
+        return match::name("transpose")(match::arg(0)(match::name("slice").bind("slice")))
+            .bind("trans");
+    }
+
+    void apply(program& p, match::matcher_result r) const
+    {
+        auto slc   = r.instructions["slice"];
+        auto trans = r.instructions["trans"];
+
+        auto input         = slc->inputs().front();
+        auto split_outputs = get_splits(input);
+        if(split_outputs.empty())
+        {
+            return;
+        }
+
+        // all transpose are the same
+        auto perm = any_cast<op::transpose>(trans->get_operator()).dims;
+        if(!std::all_of(split_outputs.begin(), split_outputs.end(), [&](auto i) {
+               assert(i->outputs().size() == 1);
+               auto tr_ins = i->outputs().front();
+               return (tr_ins->name() == "transpose") and
+                      (any_cast<op::transpose>(tr_ins->get_operator()).dims == perm);
+           }))
+        {
+            return;
+        }
+
+        // insert an transpose instruction
+        auto tr = p.insert_instruction(std::next(input), op::transpose{perm}, input);
+
+        // compute the axis in the slice
+        auto axis = any_cast<op::slice>(slc->get_operator()).axes.front();
+
+        auto it = std::find(perm.begin(), perm.end(), axis);
+        if(it == perm.end())
+        {
+            return;
```

Comment:
Can you add a test case to cover this branch?

Reply:
Looks like this is not going to happen, I just removed this check and add test case to cover other checks.

Reply:
Well you could make it an assert to help improve debugging.

Reply:
added